tensorizer/serialization.py

```diff
@@ -3700,529 +3766,152 @@ def write_tensor(
+            t.tensor.element_size()
+            * t.tensor.nelement()
+            * (not t.tensor.is_meta)
+            for t in tensors
         )
-        return tensor_endpos
-
-    @staticmethod
-    def _async_bulk_device_to_host_transfer(
-        tensors, max_read_ahead: Optional[int] = 32
-    ) -> Tuple[Iterator[torch.Tensor], Callable]:
-        """
-        Transfers CUDA tensors to host memory asynchronously in bulk.
-
-        Args:
-            tensors: The list of tensors to transfer.
-            max_read_ahead: The maximum number of tensors to queue.
-
-        Returns:
-            A tuple containing an iterator over CPU tensors,
-            and a callback to cancel the transfer early.
-        """
-        if len(tensors) < max_read_ahead:
-            transferred = queue.SimpleQueue()
-        else:
-            transferred = queue.Queue(maxsize=max_read_ahead)
-
-        tensor_sizes = [t.element_size() * t.nelement() for t in tensors]
-        staging_tensor = torch.empty(
-            (max(tensor_sizes),),
-            dtype=torch.uint8,
-            device="cpu",
-            pin_memory=True,
+        # Rough underestimate of header size
+        header_min_size = 24
+        size += header_min_size * len(tensors)
+
+        return self._header_writer_pool.submit(
+            _syscalls.try_fallocate,
+            self._fd,
+            next_pos,
+            size,
+            suppress_all_errors=True,
         )
 
-        transfer_finished = False
+    def _bulk_write(self, write_specs: Iterable[_WriteSpec], incremental=False):
+        write_specs = list(write_specs)
 
-        def _transfer():
-            nonlocal transfer_finished
-            stream = torch.cuda.Stream()
-            with torch.cuda.stream(stream):
-                # This is in a separate CUDA stream because it shouldn't
-                # affect any other GPU operations, even though each
-                # of these transfers are synchronous
-                try:
-                    for t, nbytes in zip(tensors, tensor_sizes):
-                        if transfer_finished:
-                            break
-                        staging_tensor_view = (
-                            staging_tensor.narrow(0, 0, nbytes)
-                            .view(t.dtype)
-                            .view(t.shape)
-                        )
-                        staging_tensor_view.copy_(t)
-                        new_cpu_tensor = staging_tensor_view.clone()
-                        transferred.put(
-                            new_cpu_tensor.detach(),
-                            timeout=_TIMEOUT,
-                        )
-                finally:
-                    # Sentinel
-                    transferred.put(None)
-                    transfer_finished = True
+        write_dependency: Optional[concurrent.futures.Future] = None
+        if not incremental:
+            write_dependency = self._maybe_fallocate(write_specs)
```

Comment:
Yeah, but the logic inside `_maybe_fallocate()` uses its `tensors` list to calculate total size. To work in incremental mode it'd need to be updated to account for past writes.

Reply:
https://github.com/coreweave/tensorizer/issues/150

Reply:
`fallocate` extends the file relative to the current file position, so it doesn't need anything special to account for past writes if the file position is reliable; it can just progressively extend the file.